### PR TITLE
Callback onFail returns Skygear Error object

### PR DESCRIPTION
**Pull request by howawong**
Connects #67

### CODE REVIEW

File: chat/src/main/java/io/skygear/plugins/chat/ChatContainer.java [hunk 1]
```diff
@@ -456,7 +457,7 @@ public void onSucc(@Nullable final Conversation conversation) {
                 }
 
                 if (conversation == null) {
-                    callback.onFail("Cannot find the conversation");
+                    callback.onFail(new ConversationNotFoundError(conversation.getId()));
```
Comment (carmenlau) [inline chat/src/main/java/io/skygear/plugins/chat/ChatContainer.java]: `if (conversation == null)` we should not be able to getId from the conversation?

File: chat/src/main/java/io/skygear/plugins/chat/ChatContainer.java [hunk 2]
```diff
@@ -120,7 +121,7 @@ public void onLambdaSuccess(JSONObject result){
                     public void onLambdaFail(Error error) {
 
                         if (callback != null) {
-                            callback.onFail(error.getMessage());
+                            callback.onFail(error);
```
Comment (carmenlau) [inline chat/src/main/java/io/skygear/plugins/chat/ChatContainer.java]: I think it won't be better if we could have something similar to `ConversationNotFoundError` that user can get the conversation directly? Open the discuss

File: chat/src/main/java/io/skygear/plugins/chat/ChatContainer.java (hunk 1 above)
Comment (howawong) [reply chat/src/main/java/io/skygear/plugins/chat/ChatContainer.java]: Oh my bad :(

File: chat/src/main/java/io/skygear/plugins/chat/ChatContainer.java (hunk 2 above)
Comment (howawong) [reply chat/src/main/java/io/skygear/plugins/chat/ChatContainer.java]: won't be better or would be better?

Comment (carmenlau) [reply chat/src/main/java/io/skygear/plugins/chat/ChatContainer.java]: would be better :P

**Comment by carmenlau** [comment]
The PR looks good. Only 2 questions would like to ask or discuss.

1. I think we should update the guide and api reference about the change. For guide, I believe we have to do it manually. But how about api reference?
2. About the backward compatibility issue. After we merged it, user will have to update their code to use error object if they update the SDK. We have 2 options for the handling
     1. If there are not much user using the SDK, we can leave it as it. User will have to change their code.
     2. Not sure if it is ok for the code. If we keep both `onFail(@NonNull Error error)` and `onFail(@Nullable String failReason)` in callback.

Would like to hear your thought on q2, I also put it in this week product meeting discussion.

**Comment by chpapa** [comment]
Notes after Product Meeting:

* The backward incompatibility is easy to fix, so we will assume the user will need to change it.
* API References and Guides will be updated with these error handling
* This PR will need to be updated with, handling the `distinctByParticipants` error.

**Pull request merged**